### ISSUE TITLE
Incremental rendering

### ISSUE THREAD

**Issue by fulldecent** [fulldecent/FDWaveformView#2]
For a large audio file, update the view incrementally rather than all at once

**Comment by fulldecent** [comment]
This feature conflicts slightly with the current feature of incrementally adding resolution

**Comment by scheung38** [comment]
So this is similar to the EZAudio audio recording project that could either show real time or rolling buffer display while recording audio from microphone?

**Comment by fulldecent** [comment]
No, live rendering will always be out of scope of this project (sans some hack) because we need to have the waveform in memory for zooming

**Comment by RomeHein** [comment]
Hi! I'm trying to render a waveform for a 1 hour track. Of course it takes a lot of time, so I was wondering how to implement the incremental update. Would you call your completionHandler at the end of each loop in the method "sliceAndDownsampleAsset" and then do as you suggested in your second comment? 
Thanks for your help ;)

**Comment by fulldecent** [comment]
This issue is blocked by consideration of #80. That may affect how we would be able to implement this.

**Comment by fulldecent** [comment]
This issue is no reopen for consideration.

Items include:

* Chunking the input file and only loading parts of the file into memory at a time. This will require storing multiple waveforms, something we currently do not support.

* Reporting the partial progress of a render operation as a callback.

**Comment by nuthinking** [comment]
Any progress on this feature? It would be a **massive** UX improvement. Thanks a lot!

**Comment by fulldecent** [comment]
No progress. This issue is in scope for the project. But it is "pull requests welcome" for the foreseeable future.

**Comment by fulldecent** [comment]
The latest version is published today and it has progress on this. 

https://github.com/fulldecent/FDWaveformView/blob/main/Sources/FDWaveformView/FDWaveformDataSource.swift

Here is the current data source protocol. There is a commented out function here showing a low resolution data source. We can use this in the future to quickly generate the waveform first and then generate the real wave form later. This approach allows us to get something on the screen quickly and then progress incrementally from there. 

Making this happens even for a one hour long video file will mean that we'll need to get 100 or so quick samples across the video to get its relative volume. We will have to try this to think if that is an acceptable solution or not if it is done very quickly that might be good enough while the main wave form is rendering. 

If it is not good enough, then we should try a different approach where the data source will return several times with different data sets, showing progress loading from the beginning of the file and then progressing on to the end of the file. Perhaps this will be once per second or some other frequency or perhaps it will be observable.
